crates/handlers/src/graphql/mutations/user.rs

```diff
@@ -385,4 +446,104 @@ impl UserMutations {
 
         Ok(AllowUserCrossSigningResetPayload::Allowed(user))
     }
+
+    async fn set_password(
+        &self,
+        ctx: &Context<'_>,
+        input: SetPasswordInput,
+    ) -> Result<SetPasswordPayload, async_graphql::Error> {
+        let state = ctx.state();
+        let user_id = NodeType::User.extract_ulid(&input.user_id)?;
+        let requester = ctx.requester();
+
+        if !requester.is_owner_or_admin(&UserId(user_id)) {
+            return Err(async_graphql::Error::new("Unauthorized"));
+        }
+
+        let mut policy = state.policy().await?;
+
+        let res = policy.evaluate_password(&input.new_password).await?;
+
+        if !res.valid() {
+            // TODO Expose the reason for the policy violation
+            // This involves redesigning the error handling
+            // Idea would be to expose an errors array in the response,
+            // with a list of union of different error kinds.
+            return Ok(SetPasswordPayload {
+                status: SetPasswordStatus::InvalidNewPassword,
+            });
+        }
+
+        let mut repo = state.repository().await?;
+        let Some(user) = repo.user().lookup(user_id).await? else {
+            return Ok(SetPasswordPayload {
+                status: SetPasswordStatus::NotFound,
+            });
+        };
+
+        let password_manager = state.password_manager();
+        if !requester.is_admin() {
+            // If the user isn't an admin, we:
+            // - check that password changes are enabled
+            // - check that they know their current password
+
+            if !state.site_config().password_change_allowed || !password_manager.is_enabled() {
+                return Ok(SetPasswordPayload {
+                    status: SetPasswordStatus::PasswordChangesDisabled,
+                });
+            }
+
+            let Some(active_password) = repo.user_password().active(&user).await? else {
+                // The user has no current password, so can't verify against one.
+                // TODO It may be desirable to let the user set a password without any other
+                // verification instead.
+
+                return Ok(SetPasswordPayload {
+                    status: SetPasswordStatus::NoCurrentPassword,
+                });
+            };
+
+            let Some(current_password_attempt) = input.current_password else {
+                // TODO Maybe this is the wrong error code for when you try to change
+                // your own password without supplying the current password?
+                return Ok(SetPasswordPayload {
+                    status: SetPasswordStatus::NotAllowed,
+                });
```

Comment:
Considering we're throwing a big "unauthorized" error in case you're trying to edit another user than yours, it might be okay to throw in a case where you're not calling the API the way you're supposed to?

Reply:
The only counterpoint I can think of is if you were a server administrator (who can change their own password without their current password. Not sure if this is dodgy or not.) but then your administrator privilege is taken away before you try to change your password that way.

However you could say the same thing about editing another user's password. So yes, let's just throw the big error either way I guess.